### PR TITLE
[Data] Deprecate `Reader`

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

**tl;dr** `Reader` adds unnecessary complexity, so we're removing it.

We're deprecating `Reader` and moving its methods into `Datasource`. There are several reasons why:
1. `Reader` introduces a pass-through layer.
```python
class FileBasedDatasource(Datasource):
    def create_reader(self, **kwargs):
        return _FileBasedDatasourceReader(self, **kwargs)
```
2. `Reader` subclasses contain circular dependencies.

3. `Datasource` parameters are specified with implicit (and error-prone) keywords arguments. For example:
https://github.com/ray-project/ray/blob/8504563be5302f424f6efa7ba38c03ddccb3622b/python/ray/data/datasource/csv_datasource.py#L35-L38

Going forward, you would implement `Datasource.get_read_tasks` and `Datasource.estimate_inmemory_size` directly rather than `Reader.get_read_tasks` and `Reader.estimate_inmemory_size`.

This change is part of a larger effort to clean up the `Datasource`-related interfaces. For more information, see [the design document](https://docs.google.com/document/d/1Bqhbzvxv7liwpOhyBzRVy5tOzXdy-NiMSFa-6hupr18/edit).

## Related issue number

Towards #40296 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by stephanie-wang** [comment]
Can you provide some context in the PR description?

File: python/ray/data/datasource/datasource.py [hunk 1]
```diff
@@ -27,11 +28,11 @@ class Datasource:
 
     .. note::
         Datasource instances must be serializable, since
-        :meth:`~ray.data.Datasource.create_reader` and
-        :meth:`~ray.data.Datasource.write` are called in remote tasks.
+        :meth:`~ray.data.Datasource.write` is called in remote tasks.
 
     """  # noqa: E501
 
+    @Deprecated(message="TODO")
```
Comment (stephanie-wang) [inline python/ray/data/datasource/datasource.py:35]: Do these need to be filled out?

**Review by stephanie-wang — APPROVED** [review]
Looks good, modulo the TODOs!

File: python/ray/data/datasource/datasource.py (hunk 1 above)
Comment (bveeramani) [reply python/ray/data/datasource/datasource.py]: Ah, we don't need them. I've added the `warnings.warn` message, but removed the optional `@Deprecated` message
